### ISSUE TITLE
Listen key expired despite keep-alive

### ISSUE THREAD

**Issue by edwardgonen** [JKorf/Binance.Net#796]
Hi
I'm not sure it is in your code. Maybe it is just a bug in Binance itself. But I've encountered the following problem:
After running for several hours, I've got the Listen key expired call back. My log is below. You may see that I'm sending the keep-alive message every 10 minutes. Isn't it enough? My log is right after the code fragment below. 

My keep alive code is exactly as in one of your answers:

```
        private async void KeepAliveTask()
        {
            do
            {
                try
                {
                    //keep alive. Send periodically - every 10 mins. for example
                    Logger.Instance(LogLevel.INFO, "Sending keep alive");
                    var client = new BinanceClient();
                    WebCallResult<object> reqResult = await client.FuturesUsdt.UserStream.KeepAliveUserStreamAsync(ListenKey, _cancellationToken);
                    if (!reqResult.Success)
                    {
                        Logger.Instance(LogLevel.ERROR, "Keep alive error " + reqResult.ResponseStatusCode);
                    }
                    await Task.Delay(TimeSpan.FromMinutes(10), _cancellationToken);
                }
                catch (TaskCanceledException)
                {
                    Logger.Instance(LogLevel.INFO, @"Persistence process is exiting");
                    return;
                }
            } while (!_cancellationToken.IsCancellationRequested);
        }
```

_2021-07-05 18:10:58.6966|INFO||Sending keep alive
2021-07-05 18:21:00.0322|INFO||Sending keep alive
2021-07-05 18:31:01.3656|INFO||Sending keep alive
2021-07-05 18:41:02.9946|INFO||Sending keep alive
2021-07-05 18:45:45.5889|WARN||Listen key expired
2021-07-05 18:45:45.5889|WARN||Listen key expired
2021-07-05 18:51:04.2805|INFO||Sending keep alive
2021-07-05 18:51:04.6805|ERROR||Keep alive error BadRequest_

**Comment by JKorf** [comment]
That is weird.. Seems to me you're calling the keep alive correctly, assuming the listen key is also created by the `FuturesUsdt.UserStream.StartUserStreamAsync`. 
Calling keep alive should be plenty, it's recommended to be called every 30 minutes, so 10 minutes should be fine. I see you're also checking the `KeepAliveUserStreamAsync` call for success status, so that would mean the request was successfully sent and received on the server.

Maybe try to add the ListenKey parameter value to the logging to make sure you're sending the keep alive on the same listen key as you're subscribed to. If that's the same then I'd guess this is an issue on Binance's side..

**Comment by edwardgonen** [comment]
I totally agreee it is weird. I have two servers at different locations. One (with a slower connection to binance) shows this behavior while the second one didn't. Anyway I've decided to add the following reconnection code to the OnListenKeyExpired callback. Would be great if you can give your opinion on it:

Disconnect:
`_socketClient.UnsubscribeAll().GetAwaiter().GetResult();
            WebCallResult<object> resultStopUserStream = _restClient.FuturesUsdt.UserStream.StopUserStreamAsync(ListenKey, _cancellationToken).GetAwaiter().GetResult();
            _socketClient.Dispose();
            _restClient.Dispose();`

And then full reconnect:
`BinanceClient.SetDefaultOptions(new BinanceClientOptions()
                {
                    ApiCredentials = new ApiCredentials(_apiKey, _secretKey)
                });
                BinanceSocketClient.SetDefaultOptions(new BinanceSocketClientOptions()
                {
                    ApiCredentials = new ApiCredentials(_apiKey, _secretKey)
                });

                _socketClient = new BinanceSocketClient();
                _restClient = new BinanceClient();

                WebCallResult<string> resultStartUserStream = _restClient.FuturesUsdt.UserStream.StartUserStreamAsync(_cancellationToken).GetAwaiter().GetResult();
                if (!resultStartUserStream.Success)
                {
                    resultMessage = "Failed to start user stream - " + resultStartUserStream.Data;
                    return false;
                }
                ListenKey = resultStartUserStream.Data;

                _keepAliveTask = Task.Run(KeepAliveTask, _cancellationToken);
 
                CallResult<UpdateSubscription> resultSubcriptionToUserData =
                    _socketClient.FuturesUsdt.SubscribeToUserDataUpdatesAsync(ListenKey, 
                        OnLeverageUpdate, 
                        OnMarginUpdate, 
                        OnAccountUpdate, 
                        OnOrderUpdate, 
                        OnListenKeyExpired
                    ).GetAwaiter().GetResult();
                 
                if (!resultSubcriptionToUserData.Success)
                {
                    resultMessage = "Failed to subscribe for user data changes - " + resultSubcriptionToUserData.Data.ToString();
                    return false;
                }

                resultMessage = @"Connected";
                result = true;
            }
            catch (Exception e)
            {
                resultMessage = e.ToString();
                Logger.Instance.(LogLevel.FATAL, "Failed to connect to Binance " + resultMessage);
                result = false;
            }`

**Comment by JKorf** [comment]
Seems okay to me. Is there a reason you can't use `await` but instead are using `.GetAwaiter().GetResult();`? 
Also I see in your previously posted logging that you get 
`2021-07-05 18:45:45.5889|WARN||Listen key expired` twice at the same time? Why is that?

**Comment by edwardgonen** [comment]
Thanks for your input.
Regarding:
a. Due to the async "propagation" nature, right now using await means quite a lot of changes up the link of calls. So I might do that later as for now I need to comply with existing method signature.
b. I guess I got the expiration twice. I have only one line of log in my code:
`        private void OnListenKeyExpired(DataEvent<BinanceStreamEvent> obj)
        {
            Logger.Instance(LogLevel.WARN, "Listen key expired");
...
}`

**Comment by JKorf** [comment]
Are you sure you're only subscribing to the user stream once? Maybe creating 2 socket connections on the same listen key is problematic (just speculating here)

**Comment by edwardgonen** [comment]
Just rechecked again. I'm subscribing to the user stream only once. I guess this expiration thing is something Binance-related. It happened twice on the same computer (the slow one with relatively long ping to binance). First time was when I first saw the problem. The second time I saw it in the log but my code had successfully reconnected.

So I don't know the reason for expiration nor the double log... Anyway, will monitor and update here if I see anything related.
Thanks for replies and great product.
